File: sdks/go/providers/minimax/format.go
package minimax

import (
	schema "github.com/speechswitch/client/sdks/go/generated/minimax"
	"github.com/speechswitch/client/sdks/go/runtime"
)

func audioSettings(output any, socket bool) map[string]any {
	format := "mp3"
	var rate, bitrate, channel runtime.Optional[float64]
	var cbr runtime.Optional[bool]
	switch v := output.(type) {
	case *schema.TtsRequestText77d171beOutputAsFlac:
		return audioSettings(*v, socket)
	case schema.TtsRequestText77d171beOutputAsFlac:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestText77d171beOutputAsMp3:
		return audioSettings(*v, socket)
	case schema.TtsRequestText77d171beOutputAsMp3:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestText77d171beOutputAsWav:
		return audioSettings(*v, socket)
	case schema.TtsRequestText77d171beOutputAsWav:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestTextf2dcc77eOutputAsObject:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputAsObject:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestTextf2dcc77eOutputAsMp3:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputAsMp3:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestTextf2dcc77eOutputAsMulaw:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputAsMulaw:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestTextf2dcc77eOutputAsWav6dd8e06a:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputAsWav6dd8e06a:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestTextf2dcc77eOutputAsOggOpus:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputAsOggOpus:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestTextf2dcc77eOutputAsWava066cb88:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputAsWava066cb88:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestStreamingTextaa771f19OutputAsObject:
		return audioSettings(*v, socket)
	case schema.TtsRequestStreamingTextaa771f19OutputAsObject:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestStreamingTextaa771f19OutputAsMulaw:
		return audioSettings(*v, socket)
	case schema.TtsRequestStreamingTextaa771f19OutputAsMulaw:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestStreamingTextaa771f19OutputAsWav:
		return audioSettings(*v, socket)
	case schema.TtsRequestStreamingTextaa771f19OutputAsWav:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestStreamingTextaa771f19OutputAsOggOpus:
		return audioSettings(*v, socket)
	case schema.TtsRequestStreamingTextaa771f19OutputAsOggOpus:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestStreamingTextaa771f19OutputAsMp3:
		return audioSettings(*v, socket)
	case schema.TtsRequestStreamingTextaa771f19OutputAsMp3:
		return audioSettings(v.Value, socket)
	case *schema.TtsRequestText77d171beOutputFlac:
		return audioSettings(*v, socket)
	case schema.TtsRequestText77d171beOutputFlac:
		format = v.Format.Value()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.LiteralValue())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
	case *schema.TtsRequestText77d171beOutputMp3:
		return audioSettings(*v, socket)
	case schema.TtsRequestText77d171beOutputMp3:
		format = v.Format.Value()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.LiteralValue())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
		if v.BitRateBps.Present {
			bitrate = runtime.Some(v.BitRateBps.Value.LiteralValue())
		}
		if v.ConstantBitRate.Present {
			cbr = runtime.Some(v.ConstantBitRate.Value.LiteralValue())
		}
	case *schema.TtsRequestText77d171beOutputWav:
		return audioSettings(*v, socket)
	case schema.TtsRequestText77d171beOutputWav:
		format = v.Format.Value()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.LiteralValue())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
	case *schema.TtsRequestTextf2dcc77eOutputObject:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputObject:
		format = v.Format.LiteralValue()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.LiteralValue())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
	case *schema.TtsRequestTextf2dcc77eOutputMulaw:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputMulaw:
		format = v.Format.Value()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.Value())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
	case *schema.TtsRequestTextf2dcc77eOutputWav6dd8e06a:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputWav6dd8e06a:
		format = v.Format.Value()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.Value())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
		format = "pcmu_wav"
	case *schema.TtsRequestTextf2dcc77eOutputOggOpus:
		return audioSettings(*v, socket)
	case schema.TtsRequestTextf2dcc77eOutputOggOpus:
		format = v.Format.Value()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.LiteralValue())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
	case *schema.TtsRequestStreamingText12421ea0Output:
		return audioSettings(*v, socket)
	case schema.TtsRequestStreamingText12421ea0Output:
		format = v.Format.Value()
		if v.SampleRateHz.Present {
			rate = runtime.Some(v.SampleRateHz.Value.LiteralValue())
		}
		if v.ChannelCount.Present {
			channel = runtime.Some(v.ChannelCount.Value.LiteralValue())
		}
		if v.BitRateBps.Present {
			bitrate = runtime.Some(v.BitRateBps.Value.LiteralValue())
		}
	}
	switch format {
	case "ogg_opus":
		format = "opus"
	case "mulaw":
		format = "pcmu_raw"
	}
	sampleRate := float64(32000)
	if format == "pcmu_raw" || format == "pcmu_wav" {
		sampleRate = 8000
	} else if format == "opus" {
		sampleRate = 24000
	}
	if rate.Present {
		sampleRate = rate.Value
	}
	channels := float64(1)
	if channel.Present {
		channels = channel.Value
	}
	result := map[string]any{"format": format, "sample_rate": sampleRate, "channel": channels}
	if format == "mp3" {
		bitRate := float64(128000)
		if bitrate.Present {
			bitRate = bitrate.Value
		}
		result["bitrate"] = bitRate
		if !socket {
			constant := false
			if cbr.Present {
				constant = cbr.Value
			}
			result["force_cbr"] = constant
		}
	}
	return result
}

File: sdks/rust/src/providers/minimax/format.rs
use super::settings::Map;
use crate::{generated::minimax::*, runtime::JsonValue};

#[derive(Default)]
pub(super) struct Audio {
    format: &'static str,
    rate: Option<f64>,
    channels: Option<f64>,
    bitrate: Option<f64>,
    cbr: Option<bool>,
}
pub(super) trait IntoAudio {
    fn into_audio(self) -> Audio;
}

impl IntoAudio for TtsRequestText77d171beOutput {
    fn into_audio(self) -> Audio {
        match self {
            Self::Flac(v) => v.into_audio(),
            Self::Mp3(v) => v.into_audio(),
            Self::Wav(v) => v.into_audio(),
        }
    }
}

impl IntoAudio for TtsRequestTextf2dcc77eOutput {
    fn into_audio(self) -> Audio {
        match self {
            Self::Object(v) => v.into_audio(),
            Self::Mp3(v) => v.into_audio(),
            Self::Mulaw(v) => v.into_audio(),
            Self::Wav6dd8e06a(v) => v.into_audio(),
            Self::OggOpus(v) => v.into_audio(),
            Self::Wava066cb88(v) => v.into_audio(),
        }
    }
}

impl IntoAudio for TtsRequestStreamingText12421ea0Output {
    fn into_audio(self) -> Audio {
        Audio {
            format: self.format.value(),
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: self.bit_rate_bps.map(|v| v.value()),
            cbr: None,
        }
    }
}

impl IntoAudio for TtsRequestStreamingTextaa771f19Output {
    fn into_audio(self) -> Audio {
        match self {
            Self::Object(v) => v.into_audio(),
            Self::Mulaw(v) => v.into_audio(),
            Self::Wav(v) => v.into_audio(),
            Self::OggOpus(v) => v.into_audio(),
            Self::Mp3(v) => v.into_audio(),
        }
    }
}

impl IntoAudio for TtsRequestText77d171beOutputFlac {
    fn into_audio(self) -> Audio {
        Audio {
            format: self.format.value(),
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: None,
            cbr: None,
        }
    }
}

impl IntoAudio for TtsRequestText77d171beOutputMp3 {
    fn into_audio(self) -> Audio {
        Audio {
            format: self.format.value(),
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: self.bit_rate_bps.map(|v| v.value()),
            cbr: self.constant_bit_rate.map(|v| v.value()),
        }
    }
}

impl IntoAudio for TtsRequestText77d171beOutputWav {
    fn into_audio(self) -> Audio {
        Audio {
            format: self.format.value(),
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: None,
            cbr: None,
        }
    }
}

impl IntoAudio for TtsRequestTextf2dcc77eOutputObject {
    fn into_audio(self) -> Audio {
        Audio {
            format: self.format.value(),
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: None,
            cbr: None,
        }
    }
}

impl IntoAudio for TtsRequestTextf2dcc77eOutputMulaw {
    fn into_audio(self) -> Audio {
        Audio {
            format: self.format.value(),
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: None,
            cbr: None,
        }
    }
}

impl IntoAudio for TtsRequestTextf2dcc77eOutputWav6dd8e06a {
    fn into_audio(self) -> Audio {
        Audio {
            format: "pcmu_wav",
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: None,
            cbr: None,
        }
    }
}

impl IntoAudio for TtsRequestTextf2dcc77eOutputOggOpus {
    fn into_audio(self) -> Audio {
        Audio {
            format: self.format.value(),
            rate: self.sample_rate_hz.map(|v| v.value()),
            channels: self.channel_count.map(|v| v.value()),
            bitrate: None,
            cbr: None,
        }
    }
}

impl Audio {
    pub fn wire(self, socket: bool) -> (Map, &'static str) {
        let format = match self.format {
            "" => "mp3",
            "ogg_opus" => "opus",
            "mulaw" => "pcmu_raw",
            v => v,
        };
        let rate = self.rate.unwrap_or(match format {
            "pcmu_raw" | "pcmu_wav" => 8000.0,
            "opus" => 24000.0,
            _ => 32000.0,
        });
        let mut fields = Map::from([
            ("format".into(), JsonValue::String(format.into())),
            ("sample_rate".into(), JsonValue::Number(rate)),
            (
                "channel".into(),
                JsonValue::Number(self.channels.unwrap_or(1.0)),
            ),
        ]);
        if format == "mp3" {
            fields.insert(
                "bitrate".into(),
                JsonValue::Number(self.bitrate.unwrap_or(128000.0)),
            );
            if !socket {
                fields.insert(
                    "force_cbr".into(),
                    JsonValue::Bool(self.cbr.unwrap_or(false)),
                );
            }
        }
        (fields, format)
    }
}
